### PR TITLE
Move rules settings to ESLint shared config: part 2 - check imports

### PR DESCRIPTION
**Pull request by Belco90**
Relates to #198 

Second part which includes:
- shared config to be able to set module where Testing Library utils can be imported from
- mechanism to check if Testing Library is considered as imported depending on shared config
- aggressive reporting by default

So the plugin is gonna assume that Testing Library is imported:
- if no "testing-library/module" set, then it's always considered as imported
- else, then it's considered as imported if "testing-library" import or custom module set within "testing-library/module" setting found

Why? To follow the philosophy discussed [here](https://github.com/testing-library/eslint-plugin-testing-library/issues/222#issuecomment-679592434) so by default the plugin is really aggressive to avoid missing code that should be reported. Then, if the plugin is reporting things out of the scope for the user, they can config that scope properly through shared settings.

First thing that can be configured is the module from where Testing Library is considered as imported. So if the user has its own utils reexported from `@testing-library/framework` in `test-utils` module, they can set that up so the plugin will consider just `@testing-library/framework` and `test-utils` as modules where TL utils can be imported from.

This mechanism is implemented globally through `detectTestingLibraryUtils` so rules don't need to take care of this manually since the rule creator will prevent reporting when necessary.

Next step: shared config and mechanism for test file name patterns. You'll find more comments throughout this PR.

### CODE REVIEW

File: lib/detect-testing-library-utils.ts [hunk 1]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
   TRuleListener extends TSESLint.RuleListener = TSESLint.RuleListener
 >(
   ruleCreate: (
-    context: Readonly<TSESLint.RuleContext<TMessageIds, TOptions>>,
+    context: Readonly<
+      TSESLint.RuleContext<TMessageIds, TOptions> & {
+        settings: TestingLibrarySettings;
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:23]: I can't think of a better way of typing the settings!

File: lib/detect-testing-library-utils.ts [hunk 2]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
+      }
+    >,
     optionsWithDefault: Readonly<TOptions>
-  ): TRuleListener => {
-    let isImportingTestingLibrary = false;
+  ): TSESLint.RuleListener => {
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:37]: I've replaced returned type from `TRuleListener` to `TSESLint.RuleListener` since `detectTestingLibraryUtils` modify the received rules listener to add or remove some so it can end being a different set of rule listener.

File: lib/detect-testing-library-utils.ts [hunk 3]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
-    // TODO: init here options based on shared ESLint config
+    // Init options based on shared ESLint settings
+    const customModule = context.settings['testing-library/module'];
 
-    // helpers for Testing Library detection
+    // Helpers for Testing Library detection.
     const helpers: DetectionHelpers = {
-      getIsImportingTestingLibrary() {
-        return isImportingTestingLibrary;
+      getIsTestingLibraryImported() {
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:46]: I should probably add a jsdoc here explaining the motivation behind this behavior that I wrote in this PR description, right?

File: lib/detect-testing-library-utils.ts [hunk 4]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
+      },
+      canReportErrors() {
+        return this.getIsTestingLibraryImported();
       },
     };
 
-    // instructions for Testing Library detection
+    // Instructions for Testing Library detection.
+    // `ImportDeclaration` must be first in order to know if Testing Library
+    // is imported ASAP.
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:60]: I'm still surprised this work. I can elaborate more on this but basically:

1. `detectTestingLibraryUtils` inject a detection rule listener for "ImportDeclaration" which will try to find a Testing Library import
2. before running checks from the original rule listener, `detectTestingLibraryUtils` will make sure Testing Library is imported to prevent original rule listener to be executed

I think this actually works not because the order of the rule listener definitions here, but because imports always happen first in the code linted. I just realized about this rereading my own code and try different orders here and in the concat below. I think I'm gonna remove this comment then to avoid any misunderstanding.

File: lib/detect-testing-library-utils.ts [hunk 5]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
 
     // update given rule to inject Testing Library detection
     const ruleInstructions = ruleCreate(context, optionsWithDefault, helpers);
-    const enhancedRuleInstructions = Object.assign({}, ruleInstructions);
+    const enhancedRuleInstructions: TSESLint.RuleListener = {};
+
+    // The order here is important too: detection instructions must come before
+    // than rule instructions to:
+    //  - detect Testing Library things before the rule is applied
+    //  - be able to prevent the rule about to be applied if necessary
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:87]: Second part of my previous comment. As I said, I don't finally think this order matters.

File: lib/detect-testing-library-utils.ts [hunk 6]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
+    const allKeys = new Set(
+      Object.keys(detectionInstructions).concat(Object.keys(ruleInstructions))
+    );
 
-    Object.keys(detectionInstructions).forEach((instruction) => {
-      (enhancedRuleInstructions as TSESLint.RuleListener)[instruction] = (
-        node
-      ) => {
+    allKeys.forEach((instruction) => {
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:92]: Now I iterate through **all** keys to be able to prevent original rule instructions.

File: lib/detect-testing-library-utils.ts [hunk 7]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
+      enhancedRuleInstructions[instruction] = (node) => {
         if (instruction in detectionInstructions) {
           detectionInstructions[instruction](node);
         }
 
-        if (ruleInstructions[instruction]) {
+        if (helpers.canReportErrors() && ruleInstructions[instruction]) {
```
Comment (Belco90) [inline lib/detect-testing-library-utils.ts:98]: Here is where the plugin prevents original rule instructions to be executed if conditions are not met!

File: lib/rules/no-node-access.ts [hunk 8]
```diff
@@ -25,11 +25,10 @@ export default createTestingLibraryRule<Options, MessageIds>({
   },
   defaultOptions: [],
 
-  create: (context, _, helpers) => {
+  create(context) {
     function showErrorForNodeAccess(node: TSESTree.MemberExpression) {
       isIdentifier(node.property) &&
         ALL_RETURNING_NODES.includes(node.property.name) &&
-        helpers.getIsImportingTestingLibrary() &&
```
Comment (Belco90) [inline lib/rules/no-node-access.ts:32]: This is not necessary anymore! ✨

File: tests/create-testing-library-rule.test.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
+import { createRuleTester } from './lib/test-utils';
```
Comment (Belco90) [inline tests/create-testing-library-rule.test.ts:1]: I added this test file to check both `createTestingLibraryRule` + `detectTestingLibraryUtils` together, covering as much edge cases as possible.

File: tests/fake-rule.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
+/**
```
Comment (Belco90) [inline tests/fake-rule.ts:1]: Fake rule trying to report stuff asap to be able to run the previous test file.

File: tests/lib/rules/no-node-access.test.ts [hunk 11]
```diff
@@ -51,22 +51,38 @@ ruleTester.run(RULE_NAME, rule, {
         within(signinModal).getByPlaceholderText('Username');
       `,
     },
-    {
+    /*{
```
Comment (Belco90) [inline tests/lib/rules/no-node-access.test.ts:54]: I'll reenable this test in the PR part 3 since there is where I'm gonna include the test file name pattern to report or not rules, so this will be solved then.

File: tests/lib/rules/no-node-access.test.ts [hunk 12]
```diff
@@ -51,22 +51,38 @@ ruleTester.run(RULE_NAME, rule, {
...
+      },
     },
   ],
   invalid: [
+    {
+      code: `
+      // case: without importing TL (aggressive reporting)
+      const closestButton = document.getElementById('submit-btn')
+      expect(closestButton).toBeInTheDocument();
+      `,
```
Comment (Belco90) [inline tests/lib/rules/no-node-access.test.ts:83]: This was a valid case before but it isn't anymore. I'll find more like this in other rules.

File: lib/detect-testing-library-utils.ts [hunk 13]
```diff
@@ -13,48 +18,84 @@ export function detectTestingLibraryUtils<
...
-        return isImportingTestingLibrary;
+      getIsTestingLibraryImported() {
+        if (!customModule) {
+          return true;
+        }
+
+        return isImportingTestingLibraryModule || isImportingCustomModule;
+      },
+      canReportErrors() {
+        return this.getIsTestingLibraryImported();
```
Comment (gndelia) [inline lib/detect-testing-library-utils.ts:54]: it's been a long time since I've seen a `this` 😄

File: tests/fake-rule.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+          node,
+          messageId: 'fakeError',
+        });
+      }
+    };
+
+    return {
+      'CallExpression Identifier': reportRenderIdentifier,
+    };
+  },
```
Comment (gndelia) [inline tests/fake-rule.ts:41]: Can we add to this rule a bunch of `Import` keys? This way we could test that the `Import` key we check does not override any `Import` the user could've defined in their rules. For instance, an example I could think of is a rule that verifies that some deprecated methods are not imported (like the `prefer-wait-for` rule)

File: lib/detect-testing-library-utils.ts (hunk 7 above)
Comment (gndelia) [reply lib/detect-testing-library-utils.ts:98]: I'm a bit lost what's the difference between the method `canReportErrors` and `getIsTestingLibraryImported` - given the first one just calls the second

Comment (gndelia) [reply lib/detect-testing-library-utils.ts:98]: in addition to that, if I understand correctly, we would just execute the rules if TL is imported, otherwise, nothing gets executed - is that correct?

Comment (Belco90) [reply lib/detect-testing-library-utils.ts:98]: I'm anticipating next PR with `canReportErrors` so it warps all necessary checks to know if rules should report errors or not. That's just knowing if Testing Library is imported from now, but in the next PR I'll add the shared option to customize the file names pattern which should be checked by regexp, so the user can set to just report from this plugin on files named `*.test.js`. This is interesting when you have different kind of tests files (.test.js for Testing Library and .spec.js for Cypress) or just to reduce the scope of the plugin. So in the next PR `canReportErrors` will check more things and will be in charge of wrap all necessary checks in the future if others needed.

> in addition to that, if I understand correctly, we would just execute the rules if TL is imported, otherwise, nothing gets executed - is that correct?

Yes, but by default TL is always considered as imported due to the "aggressive reporting" mode.

File: tests/fake-rule.ts (hunk 14 above)
Comment (Belco90) [reply tests/fake-rule.ts:41]: Sure, I can do that.

**Review by gndelia — APPROVED** [review]
Looking great 🚀

**Comment by Belco90** [comment]
Thanks for your review, merging now! Can't wait to have the rest of the shared config, it's gonna make rules implementation much easier.

My idea is to create the PR for the option to customize the test files name pattern next week, reenabling the commented test out in this PR and updating the login for `canReportErrors`. Then I'll update couple of rules to use the new rule creator.

After that I'll add the last shared config for defining custom `render` methods, which will be the most complex one.
